pkg/controller/podautoscaler/horizontal_test.go

```diff
@@ -4103,3 +4103,22 @@ func TestNoScaleDownOneMetricEmpty(t *testing.T) {
 	tc.testEMClient = testEMClient
 	tc.runTest(t)
 }
+
+func TestConditionZeroDesiredReplicasWithInvalidMetric(t *testing.T) {
+	tc := testCase{
+		minReplicas:             1,
+		maxReplicas:             4,
+		specReplicas:            1,
+		statusReplicas:          1,
+		expectedDesiredReplicas: 1,
```

Comment:
This test does not fail without your changes (doesn't verify the fix) because `status.desiredReplicas` is set to `tc.specReplicas` here: https://github.com/kubernetes/kubernetes/blob/1a9ae3454993b2362503db3efe1d7e48307bf3ff/pkg/controller/podautoscaler/horizontal_test.go#L227

See here it is passing when it should not:
```
josephburnett@josephburnett:~/go/src/k8s.io/kubernetes$ git status
On branch hotfix/hpa-desired-replicas
Your branch is up to date with 'hysyeah/hotfix/hpa-desired-replicas'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pkg/controller/podautoscaler/horizontal.go

no changes added to commit (use "git add" and/or "git commit -a")
josephburnett@josephburnett:~/go/src/k8s.io/kubernetes$ git diff
diff --git a/pkg/controller/podautoscaler/horizontal.go b/pkg/controller/podautoscaler/horizontal.go
index cba9436c047..9252085fabd 100644
--- a/pkg/controller/podautoscaler/horizontal.go
+++ b/pkg/controller/podautoscaler/horizontal.go
@@ -652,9 +652,10 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
                var metricTimestamp time.Time
                metricDesiredReplicas, metricName, metricStatuses, metricTimestamp, err = a.computeReplicasForMetrics(hpa, scale, hpa.Spec.Metrics)
                if err != nil {
-                       // set desiredReplicas to scale.Spec.Replicas when can't compute metrics
-                       desiredReplicas = scale.Spec.Replicas
-                       a.setStatus(hpa, currentReplicas, desiredReplicas, hpa.Status.CurrentMetrics, false)
+                       a.setCurrentReplicasInStatus(hpa, currentReplicas)
+                       // // set desiredReplicas to scale.Spec.Replicas when can't compute metrics
+                       // desiredReplicas = scale.Spec.Replicas
+                       // a.setStatus(hpa, currentReplicas, desiredReplicas, hpa.Status.CurrentMetrics, false)
                        if err := a.updateStatusIfNeeded(hpaStatusOriginal, hpa); err != nil {
                                utilruntime.HandleError(err)
                        }
josephburnett@josephburnett:~/go/src/k8s.io/kubernetes$ go test ./pkg/controller/podautoscaler/...
ok      k8s.io/kubernetes/pkg/controller/podautoscaler  (cached)
?       k8s.io/kubernetes/pkg/controller/podautoscaler/config   [no test files]
?       k8s.io/kubernetes/pkg/controller/podautoscaler/config/v1alpha1  [no test files]
ok      k8s.io/kubernetes/pkg/controller/podautoscaler/metrics  (cached)
```

Can you find a clean way to unit test this?  Maybe you could prepend a reactor to give a different scale subresource `spec.replicas` (https://github.com/kubernetes/kubernetes/blob/1a9ae3454993b2362503db3efe1d7e48307bf3ff/pkg/controller/podautoscaler/horizontal_test.go#L2339) or add a parameter to the test `initialStatusReplicas *int32` which is optional that you can set to zero.

Reply:
@josephburnett  I add a field `initialStatusReplicas` in `testCase` and change `DesiredReplicas: tc.specReplicas` to `DesiredReplicas: tc.initialStatusReplicas` . By default, `DesiredReplicas =  tc.initialStatusReplicas`, you alse can use `tc.initialStatusReplicas = tc.specReplicas` to set  `DesiredReplicas=c.specReplicas` . or you can use option method to set tc.initialStatusReplicas value.  What advice do you have?